npm/networkPolicyController.go

```diff
@@ -0,0 +1,492 @@
+func (c *networkPolicyController) syncAddAndUpdateNetPol(netPolObj *networkingv1.NetworkPolicy) error {
+	// This timer measures execution time to run this function regardless of success or failure cases
+	timer := metrics.StartNewTimer()
+	defer timer.StopAndRecord(metrics.AddPolicyExecTime)
+
+	var err error
+	netpolKey, err := cache.MetaNamespaceKeyFunc(netPolObj)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: while running MetaNamespaceKeyFunc err: %s", err)
+	}
+
+	// Start reconciling loop to eventually meet cached states against the desired states from network policy.
+	// #1 If a new network policy is created, the network policy is not in RawNPMap.
+	// start translating policy and install translated ipset and iptables rules into kernel
+	// #2 If a network policy with <ns>-<netpol namespace>-<netpol name> is applied before and two network policy are the same object (same UID),
+	// first delete the applied network policy, then start translating policy and install translated ipset and iptables rules into kernel
+	// #3 If a network policy with <ns>-<netpol namespace>-<netpol name> is applied before and two network policy are the different object (different UID) due to missing some events for the old object
+	// first delete the applied network policy, then start translating policy and install translated ipset and iptables rules into kernel
+	// To deal with all three cases, we first delete network policy if possible, then install translated rules into kernel.
+	// (TODO): can optimize logic more to reduce computations. For example, apply only difference if possible like podController
+
+	// Do not need to clean up default Azure NPM chain in deleteNetworkPolicy function, if network policy object is applied soon.
+	// So, avoid extra overhead to install default Azure NPM chain in initializeDefaultAzureNpmChain function.
+	// To achieve it, use flag unSafeToCleanUpAzureNpmChain to indicate that the default Azure NPM chain cannot be deleted.
+	// delete existing network policy
+	cachedNetPolKey := util.GetNSNameWithPrefix(netpolKey)
+	err = c.cleanUpNetworkPolicy(cachedNetPolKey, unSafeToCleanUpAzureNpmChain)
+	if err != nil {
+		return fmt.Errorf("[syncAddAndUpdateNetPol] Error: failed to deleteNetworkPolicy due to %s", err)
+	}
+
+	// Install this default rules for kube-system and azure-npm chains if they are not initilized.
+	// Execute initializeDefaultAzureNpmChain function first before actually starting processing network policy object.
+	if err = c.initializeDefaultAzureNpmChain(); err != nil {
+		return fmt.Errorf("[syncNetPol] Error: due to %v", err)
+	}
+
+	// Cache network object first before applying ipsets and iptables.
+	// If error happens while applying ipsets and iptables,
+	// the key is re-queued in workqueue and process this function again, which eventually meets desired states of network policy
+	c.npMgr.RawNpMap[cachedNetPolKey] = netPolObj
```

Comment:
We will need a deepcopy of this here ! we are not removing references from the sharedinformer cache

Reply:
I think current code is better in terms of memory and computations. If we use deepcopy network policyhere, we will always have two network policy objects and increase computations for deepcopy operation. But, if we use reference of network policy from sharedinformer cache, it will be either one or two objects eventually and no computation overhead for deepcopy.